lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -822,95 +822,139 @@ computeBinary(ExprSolution lhs, ExprSolution rhs,
-            return computeBinary(lhs, rhs, [](int32_t lhs, int32_t rhs) {
-              return std::min(lhs, rhs);
-            });
-          })
-          .Default([](auto) {
-            return ExprSolution{std::nullopt, false};
-          });
-
-  // Memoize the result.
-  if (solution.first && !solution.second)
-    expr->solution = *solution.first;
+            return setSolution(solution);
+          }
 
-  // Produce some useful debug prints.
-  LLVM_DEBUG({
-    if (!isa<KnownExpr>(expr)) {
-      if (solution.first)
-        llvm::dbgs().indent(indent * 2)
-            << "= Solved " << *expr << " = " << *solution.first;
-      else
-        llvm::dbgs().indent(indent * 2) << "= Skipped " << *expr;
-      llvm::dbgs() << " (" << (solution.second ? "cycle broken" : "unique")
-                   << ")\n";
-    }
-  });
+          // Unconstrained variables produce no solution.
+          if (!expr->constraint)
+            return setSolution(ExprSolution{std::nullopt, false});
+          // Return no solution for recursions in the variables. This is sane
+          // and will cause the expression to be ignored when computing the
+          // parent, e.g. `a >= max(a, 1)` will become just `a >= 1`.
+          if (!seenVars.insert(expr).second)
+            return setSolution(ExprSolution{std::nullopt, true});
+
+          worklist.push_back({expr->constraint, indent + 1});
+          if (expr->upperBound)
+            worklist.push_back({expr->upperBound, indent + 1});
+        })
+        .Case<IdExpr>([&](auto *expr) {
+          if (solvedExprs.contains(expr->arg))
+            return setSolution(solvedExprs[expr->arg]);
+          worklist.push_back({expr->arg, indent + 1});
+        })
+        .Case<PowExpr>([&](auto *expr) {
+          if (solvedExprs.contains(expr->arg))
+            return setSolution(computeUnary(
+                solvedExprs[expr->arg], [](int32_t arg) { return 1 << arg; }));
+
+          worklist.push_back({expr->arg, indent + 1});
+        })
+        .Case<AddExpr>([&](auto *expr) {
+          if (solvedExprs.contains(expr->lhs()))
+            return setSolution(computeBinary(
+                solvedExprs[expr->lhs()], solvedExprs[expr->rhs()],
+                [](int32_t lhs, int32_t rhs) { return lhs + rhs; }));
+
+          worklist.push_back({expr->lhs(), indent + 1});
+          worklist.push_back({expr->rhs(), indent + 1});
+        })
+        .Case<MaxExpr>([&](auto *expr) {
+          if (solvedExprs.contains(expr->lhs()))
+            return setSolution(computeBinary(
+                solvedExprs[expr->lhs()], solvedExprs[expr->rhs()],
+                [](int32_t lhs, int32_t rhs) { return std::max(lhs, rhs); }));
+
+          worklist.push_back({expr->lhs(), indent + 1});
+          worklist.push_back({expr->rhs(), indent + 1});
+        })
+        .Case<MinExpr>([&](auto *expr) {
+          if (solvedExprs.contains(expr->lhs()))
```

Comment:
Is it necessary to check `expr->rhs()` as well? I think it's necessary to check if expressions are shared across multiple expressions.